### ISSUE TITLE
pagurl tag doesn't work within StreamField

### ISSUE THREAD

**Issue by gmeans** [torchbox/wagtail#1743]
After adding a PageChooserBlock to a stream field and then trying to use the pageurl template tag, I received the following error:

KeyError at <url>
'request'

I stepped through the code and it looks like the block doesn't get the full context passed to it, so request isn't available.

**Comment by gasman** [comment]
This is a consequence of the way StreamField is designed - at some point, you have to render a value within a `{{ ... }}` tag (either the StreamField value itself, e.g. `{{ self.body }}` or a child of it, e.g. `{% for block in self.body %}{{ block }}{% endfor %}`) - and at that point you're rendering a string representation of the value. That value doesn't know anything about the template context it's being rendered in - indeed, there may not be a template context at all, since you could easily retrieve the same value by digging around on the `./manage.py shell` command line. It just so happens that the string representation of that value is (usually) produced by rendering a template, but that's an entirely separate mechanism to the request/response cycle.

Having said that, you're not the first person to encounter this, and the `{% pageurl %}` tag is a particularly strong example of why you might want to access the outer template context within a StreamField's rendering. A common pattern throughout Wagtail's architecture is "objects that know their own HTML rendering (and usually generate it with a template)" - StreamField values being one of those - so perhaps this calls for a new way of handling those, which *does* preserve the template context. We'd have to change our template-writing conventions a bit, since the `{{ ... }}` tag effectively creates a barrier that the template context can't cross. One idea, then:

* Objects that use templates to render themselves - such as StreamField values - implement an `__str__` method that takes an optional `context` argument. If provided, they render templates using that context; if not, they use a blank context.
* Where we would currently write `{{ self.body }}`, we start writing `{% render self.body %}` instead, to denote that we expect to get an HTML rendering back. `render` here would be a custom tag provided by `wagtailcore_tags`. (I'm not keen on the name 'render' here - it's horribly generic. But it's doing something horribly generic, so maybe that's unavoidable :-) )
* The `render` tag would attempt to call `__str__(context)` on the value, and if that fails, fall back to the standard `__str__()` instead.

One quick and easy fix to the immediate problem would be to make `{% pageurl %}` work correctly in the absence of a `request` variable. We check `request` to determine whether the URL is on a different site to the current page, and produce a local URL if not - but it would be simple enough to define alternative behaviour for when `request` isn't available. However, I wouldn't want that to be the only fix we implement here, as it's sweeping the bigger problem under the carpet...

**Comment by gmeans** [comment]
I think giving the StreamField blocks access to the page context would be a great addition. Logically, for me at least, the StreamField (and in effect the blocks included) are a part of that page instance. 

The addition you are proposing seems really sound considering Django's template rendering. Maybe call the tag {% with_context self.body %} if render is too generic.

Thanks for the great explanation!

**Comment by curaloucura** [comment]
I appreciate the responses but I'm confused if there's any real workaround to this issue. I'm guessing the easiest to do until a final decision on the approach is to copy and paste the pageurl to another module and avoid the request check, since I know I'm already under the same site?

**Comment by gasman** [comment]
@curaloucura The easiest workaround is to use `{{ my_page.url }}` instead of `{% pageurl my_page %}` - the pageurl tag is only really needed to ensure correct behaviour when linking between different sites on one Wagtail instance.

**Comment by gasman** [comment]
> Where we would currently write `{{ self.body }}`, we start writing `{% render self.body %}` instead, to denote that we expect to get an HTML rendering back. render here would be a custom tag provided by wagtailcore_tags. (I'm not keen on the name 'render' here - it's horribly generic. But it's doing something horribly generic, so maybe that's unavoidable :-) )

Having thought about this some more, I think `{% include_block %}` works nicely as a possible name for this tag. From the template author's point of view, the effect is somewhat similar to `{% include %}`, in that the context variables are passed on to the inner template (by default) - only instead of passing a template filename directly, you're passing a block object that decides which template to use. We can even borrow some of the extra semantics of the `{% include %}` tag...

    {% include_block advert with color="purple" %}

    {% include_block advert with page=page user=request.user color="purple" only %}

**Comment by gmeans** [comment]
We just did our second Wagtail project and ran into a lot of challenges not having the page context available in StreamField blocks. In the project we used a Custom Auth Backend that populated a user profile. We then wanted some of these profile variables in custom blocks we created.

We had a bad time :(

In the end it was easier to populate a JS variable with the profile values we wanted in the page, and then use JS to populate the value in the block template.

In this instance it would have been great to simply state:

```
{% with_context self.body %} 
```

And then have context available (so we could get anything we wanted provided by Django's context processors).

Considering that scenario, how would include_block apply?

Would it be in the page? So our implementation would be:

```
{% include_block self.body with <vars> %}
```

If that's the case then I think I'd prefer to simply have the context as opposed to having to specify each variable.

Maybe this scenario out out of scope for the original issue though.

**Comment by gasman** [comment]
@gmeans `{% include_block self.body %}` by itself would pass on the entire context of the parent template - anything passed in `with <vars>` would be in addition to that.

**Comment by gmeans** [comment]
Is an instance of StreamField actually a block object? Just playing devil's advocate but what would make include_block better than say, streamfield. Implementation could look like:

```
{% streamfield self.body %}
```

Which I think falls in line w/ how its necessary to use a filter on richtext fields.

I may not have a good understanding of the architecture though. In my mind, a StreamField is a list of blocks. So saying `include_block` doesn't really click and tell me I'm passing vars that are now available to my StreamField blocks.

**Comment by gasman** [comment]
> Is an instance of StreamField actually a block object?

Yes (give or take some technicalities that aren't important here). The top level object within a StreamField is a StreamBlock, which consists of a list of child blocks, which may in turn contain child blocks of their own (as in the case of a StructBlock - indeed, it's entirely possible to nest a StreamBlock within a StreamBlock). The reason `{% streamfield %}` isn't really suitable as a tag name is that the tag could be used at any level down that chain, not just on the streamfield as a whole:

    {% include_block self.body %}

or

    <ul>
        {% for block in self.body %}
            <li>{% include_block block %}</li>
        {% endfor %}
    </ul>

or

    # models.py
    class PersonBlock(blocks.StructBlock):
        name = blocks.CharBlock()
        photo = blocks.ImageChooserBlock()
        life_story = SomeExtremelyComplexBlock()

        class Meta:
            template = 'blocks/person.html'

    # blocks/person.html
    <h1>{{ value.name }}</h1>
    {% include_block value.photo %}
    {% include_block value.life_story %}


> So saying `include_block` doesn't really click and tell me I'm passing vars that are now available to my StreamField blocks.

I don't think developers *should* think of it as a mechanism for passing variables to the block - you're only thinking of it that way because you've been bitten by the current implementation, which *doesn't* pass variables. :-)

What I'm proposing is that `{% include_block foo %}` becomes the standard convention for inserting non-trivial block content (basically, anything that outputs HTML of its own, as opposed to a basic CharBlock), not just a variant to be used on the occasions you want to pass variables. I believe this whole problem comes about because developers are *already* viewing `{{ page.body }}` as a kind of `{% include %}`-like mechanism, which is why they're surprised when they find that `request.user` or `page` doesn't work inside the block template. By making this change, we'd be bringing the reality in line with what developers think it is, rather than using `{{ ... }}` tags for something they weren't designed for.

**Comment by gmeans** [comment]
Ah I get it now. Thanks for the great explainer.

I'm looking forward to seeing it roll out.

**Comment by mgd020** [comment]
Workaround: https://gist.github.com/mgd020/d841f8016024707b11b490b82ec648f0

**Comment by m1kola** [comment]
This issue and issue #2469 were partially solved in #2786. But the feature still needs some work (see #2823 and #2824).

**Comment by gasman** [comment]
Closing this now, as the specific issue described here is fixed by #2786 (even if there is further work to do on the feature as a whole).
